File: ruby/mergesort.rb
class Mergesort
  def self.sort(arr)
    if arr.size <= 1
      arr
    else
      half = arr.size / 2
      first_half = sort(arr[0...half])
      second_half = sort(arr[half..-1])
      merge(first_half, second_half)
    end
  end

  def self.merge(arr_one, arr_two)
    sorted = []
    while (arr_one.length > 0 && arr_two.length > 0)
      if arr_one[0] < arr_two[0]
        sorted << arr_one.slice!(0)
      else
        sorted << arr_two.slice!(0)
      end
    end
      # there's leftovers in one of them
    if arr_one.length > arr_two.length
      sorted.concat(arr_one)
    else
      sorted.concat(arr_two)
    end
    sorted
  end
end

File: c/mergesort.c
#include <stdio.h>

int * mergesort(int *arr){
  int * sorted;
  int length = sizeof(arr) / sizeof(int)
    if (length <= 1){
      return arr;
    } else {
      int * first_half = mergesort(arr + (length / 2) - 1);
      int * second_half = mergesort(arr + ());
      for (int = i; i < length / 2; i++){
        first = first_half[i];
        second = second_half[i];
        if (first < second){
          sorted[i] = first;
          sorted[i + 1] = second;
        } else {
          sorted[i] = second;
          sorted[i + 1] = first;
        }
      }

    }
  return sorted;
}

int main(int argc, char** argv){
  int nums[4];
  for (int i = 0; i < 4; i++){
    int arg;
    printf("Give me a number:\n");
    scanf("%d", &arg);
    nums[i] = arg;
  }
  return 0;
}
